tests/Validator.test.ts

```diff
@@ -0,0 +1,159 @@
+        expect(mockInfoSpy).toHaveBeenCalled();
+      });
+    },
+  );
+
+  describe.each([['bad-prolog.xml', 'unclosed-child', 'unclosed-root']])(
+    'given a problematic file',
+    (fileName) => {
+      let mockErrorSpy: any;
+      let mockInfoSpy: any;
+      let mockFatalSpy: any;
+
+      beforeAll(() => {
+        mockErrorSpy = jest.spyOn(Logger, 'error');
+        mockErrorSpy.mockImplementation(() => {});
+        mockFatalSpy = jest.spyOn(Logger, 'fatal');
+        mockFatalSpy.mockImplementation(() => {});
+        mockInfoSpy = jest.spyOn(Logger, 'info');
+        mockInfoSpy.mockImplementation(() => {});
+
+        const xmlFile = resolve(__dirname, 'files', fileName);
+        validateFile(xmlFile);
+      });
+
+      afterAll(() => {
+        jest.restoreAllMocks();
+      });
+
+      it('does not report any (regular) errors', () => {
+        expect(mockErrorSpy).not.toHaveBeenCalled();
+      });
+
+      it('does report a fatal error', () => {
+        expect(mockFatalSpy).toHaveBeenCalled();
+      });
+
+      it('logs to the info console', () => {
+        expect(mockInfoSpy).not.toHaveBeenCalled();
+      });
+    },
+  );
+});
+
+describe('when validating XML file against its schema', () => {
+  describe.each([['valid-without-prolog.xml', 'valid-with-prolog.xml']])(
+    'given a well-formed file',
+    (fileName) => {
+      let mockErrorSpy: any;
+      let mockInfoSpy: any;
+      let mockFatalSpy: any;
+
+      beforeAll(() => {
+        mockErrorSpy = jest.spyOn(Logger, 'error');
+        mockErrorSpy.mockImplementation(() => {});
+        mockFatalSpy = jest.spyOn(Logger, 'fatal');
+        mockFatalSpy.mockImplementation(() => {});
+        mockInfoSpy = jest.spyOn(Logger, 'info');
+        mockInfoSpy.mockImplementation(() => {});
+
+        const xmlFile = resolve(__dirname, 'files', fileName);
+        const xsdFile = resolve(__dirname, 'files', 'sport.xsd');
+        validateFile(xmlFile, xsdFile);
+      });
+
+      afterAll(() => {
+        jest.restoreAllMocks();
+      });
+
+      it('does not report any (regular) errors', () => {
+        expect(mockErrorSpy).not.toHaveBeenCalled();
+      });
+
+      it('does not report any fatal errors', () => {
+        expect(mockFatalSpy).not.toHaveBeenCalled();
+      });
+
+      it('logs to the info console', () => {
+        expect(mockInfoSpy).toHaveBeenCalled();
+      });
+    },
+  );
+
+  describe.each([['bad-child-element.xml', 'wrong-root-element.xml']])(
+    'given a non-conforming file',
+    (fileName) => {
+      let mockErrorSpy: any;
+      let mockInfoSpy: any;
+      let mockFatalSpy: any;
+
+      beforeAll(() => {
+        mockErrorSpy = jest.spyOn(Logger, 'error');
+        mockErrorSpy.mockImplementation(() => {});
+        mockFatalSpy = jest.spyOn(Logger, 'fatal');
+        mockFatalSpy.mockImplementation(() => {});
+        mockInfoSpy = jest.spyOn(Logger, 'info');
+        mockInfoSpy.mockImplementation(() => {});
```

Comment:
These spies seem to work some of the time, but I still get unexpected errors with them, as if the code were calling the real function instead of the spy.

Reply:
Oh yes, that's a funny behavior of `spyOn` compared to other mock libraries. See the "Note" at https://jestjs.io/docs/jest-object#jestspyonobject-methodname